postgres/datadog_checks/postgres/connections.py

```diff
@@ -92,8 +127,51 @@ def close_all_connections(self):
                     success = False
         return success
 
-    def _terminate_connection_unsafe(self, dbname):
-        db, _ = self._conns.pop(dbname, ConnectionWithTTL(None, None))
+    def done(self, dbname: str) -> None:
+        """
+        Mark a connection as done being used, so it can be evicted from the pool.
+        This function does not evict connections from the pool; it just marks them
+        as inactive.
+        done() can only be called on a connection in the same thread that the connection
+        was made.
+        """
+        with self._mu:
+            if self._conns[dbname].thread != threading.current_thread():
+                raise RuntimeError(
+                    "Cannot call done() for this dbname on this thread. Done() can only be called \
+                                   from the same thread the connection was made."
+                )
+
+            self._conns[dbname].active = False
+
+    def evict_lru(self) -> str:
+        """
+        Evict and close the inactive connection which was least recently used.
+        Return the dbname connection that was evicted.
+        """
+        with self._mu:
+            conns_list = dict(self._conns)
+            while True:
+                if not conns_list:
+                    break
+
+                eviction_candidate = self._get_lru(conns_list)
+                if self._conns[eviction_candidate].active:
```

Comment:
Did you mean
```suggestion
                if not self._conns[eviction_candidate].active:
```

otherwise I don't know if I follow the logic of deleting active connections. Can you explain?

Reply:
oh, the code isn't clear. 
I am copying the pool's connections list at the beginning of this function, and deleting connections from the copy every time they're eliminated as an eviction candidate. so, if a connection is active, it should be removed from the candidates list. maybe I should renamed that list to `candidates_list` for clarity

Reply:
Ohhhh ok that clarifies things a lot!

My other concern is moot then because I now understand how you are only terminating inactive conns. I *think* the logic can be cleaner and clearer by inverting the operations. I.e. it is currently:

* copy the list
* delete all active conns from the list
* terminate the min active connection

I would suggest:

* sorting the list on last used
* iterate until you find the first non-active
* terminate that connection and return it

This will be much more efficient algorithm too.

Reply:
oh wow that is so much better. I will do that! thank you!